keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1549 @@
+    AllocationMode AllocationMode
+}
+
+type AllocationMode string
+
+const (
+     AllocationModeImmediate AllocationMode = “Immediate”
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
+   // Explains what the current status of the claim is and
+   // determines which component needs to do something.
+   Phase ResourceClaimPhase
+
+   // A copy of the driver name from the ResourceClass at
+   // the time when allocation started. Drivers can
+   // filter claims by this field. It's also necessary to
+   // support deallocation when the class gets deleted before
+   // a claim.
+   DriverName string
+
+   // When allocation is delayed, the scheduler must set
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
+
+   // When allocation is delayed, and the scheduler needs to
+   // decide on which node a Pod should run, it will first
+   // ask the driver on which nodes the resource might be
+   // made available. To trigger that check, the scheduler
+   // provides the names of nodes which might be suitable
+   // for the Pod. Will be updated periodically until
+   // the claim is allocated.
+   PotentialNodes []string
+
+   // A change of the node candidates triggers a check
+   // on which nodes the resource could be made available.
+   // Unsuitable nodes will be ignored when selecting
+   // a node. All other nodes are potential candidates,
+   // either because no information is available yet
+   // or because allocation is expected to succeed.
+   //
+   // This can change, so the driver must refresh
+   // this information periodically until a node gets
+   // selected by the scheduler.
+   UnsuitableNodes []string
+
+   // An allocated resource is available on nodes that match this
+   // selector. If nil, the resource is available everywhere.
+   AvailableOnNodes *corev1.NodeSelector
+
+   // Arbitrary data returned by the driver after a successful allocation.
+   // This data is passed to the driver for all operations involving
+   // the allocated resource. This is opaque for Kubernetes.
+   // Driver documentation may explain to users how to interpret
+   // this data if needed.
+   //
+   // The attributes must be sufficient to deallocate the resource
+   // because the ResourceClass might not be available anymore
+   // when deallocation starts.
+   Attributes map[string]string
+
+    // UsersLimit determines how many entities are allowed to use this resource
+    // at the same time. The default is 1. -1 enables the usage by an unlimited number
+   // of users. Individual containers in a pod are not counted as users, only the Pod
+   // is.
+   UserLimit int
```

Comment:
@sftim asked about using ResourceClaim as a mechanism for resource allocation also without ever using the resource in a Pod. It wasn't our original intention, but it might make sense. To cover that case, the field is for counting "users" and not specifically "pods".

Reply:
I added that explanation to the summary.

Reply:
Even so, I also think "user" has too much baggage and we could probably come up with a better name for this field (though I admittedly can't think of anything better off the top of my head). Maybe `MaxReservations` to match the `ReservedFor` field below?